src/hierarchies/updaters/semi_conjugate_updater.h

```diff
@@ -0,0 +1,65 @@
+#ifndef BAYESMIX_HIERARCHIES_UPDATERS_SEMI_CONJUGATE_UPDATER_H_
+#define BAYESMIX_HIERARCHIES_UPDATERS_SEMI_CONJUGATE_UPDATER_H_
+
+#include <utility>
+
+#include "abstract_updater.h"
+#include "src/hierarchies/likelihoods/abstract_likelihood.h"
+#include "src/hierarchies/priors/abstract_prior_model.h"
+
+template <class Likelihood, class PriorModel>
+class SemiConjugateUpdater : public AbstractUpdater {
+ public:
+  SemiConjugateUpdater() = default;
+
+  ~SemiConjugateUpdater() = default;
+
+  void draw(AbstractLikelihood& like, AbstractPriorModel& prior,
+            bool update_params) override;
+
+  void save_posterior_hypers(ProtoHypersPtr post_hypers_) override;
+
+ protected:
+  Likelihood& downcast_likelihood(AbstractLikelihood& like_);
+  PriorModel& downcast_prior(AbstractPriorModel& prior_);
+  ProtoHypersPtr post_hypers = std::make_shared<ProtoHypers>();
+};
+
+// Methods' definitions
+template <class Likelihood, class PriorModel>
+Likelihood& SemiConjugateUpdater<Likelihood, PriorModel>::downcast_likelihood(
+    AbstractLikelihood& like_) {
+  return static_cast<Likelihood&>(like_);
+}
+
+template <class Likelihood, class PriorModel>
+PriorModel& SemiConjugateUpdater<Likelihood, PriorModel>::downcast_prior(
+    AbstractPriorModel& prior_) {
+  return static_cast<PriorModel&>(prior_);
+}
+
+template <class Likelihood, class PriorModel>
+void SemiConjugateUpdater<Likelihood, PriorModel>::draw(
+    AbstractLikelihood& like, AbstractPriorModel& prior, bool update_params) {
+  // Likelihood and PriorModel downcast
+  auto& likecast = downcast_likelihood(like);
+  auto& priorcast = downcast_prior(prior);
+  // Sample from the full conditional of a semi-conjugate hierarchy
+  bool set_card = true; /*, use_post_hypers=true;*/
+  if (likecast.get_card() == 0) {
+    likecast.set_state_from_proto(*priorcast.sample(), !set_card);
+  } else {
+    auto post_params = compute_posterior_hypers(likecast, priorcast);
+    likecast.set_state_from_proto(*priorcast.sample(post_params), !set_card);
+    if (update_params) save_posterior_hypers(post_params);
```

Comment:
Since this is a template class, I think we could avoid having the back-and-forth conversion to protos.

In fact, we could have that the `Updater::compute_posterior_hypers` return an object in the prior/hyperparams namespace. 

Then every PriorModel implements the method

```
SomeState sample(const SomeHypers& params);
```

In the BasePriorModel we provide a unique function with signature
```
std::shared_ptr<google::protobuf::Message> sample(
       ProtoHypersPtr hier_hypers = nullptr) {
   auto params static_cast<Derived*>(this)->hypers_from_proto(hier_hypers);
   auto out = params static_cast<Derived*>(this)->draw(params);
   return out.to_proto();
}
```
the `hypers_from_proto` is basically the same method as `set_hypers_from_proto` but it returns the actual object (or a pointer or whatever is easier).

What do you think?

Reply:
I don't think this is needed right now, but it would make code easier to read. Maybe once we merge the PR we can open an issue and do this when we get time

Reply:
This is a huge +1 for me!